src/main/kotlin/platform/architectury/version/ArchitecturyVersion.kt

```diff
@@ -0,0 +1,144 @@
+/*
+ * Minecraft Dev for IntelliJ
+ *
+ * https://minecraftdev.org
+ *
+ * Copyright (c) 2021 minecraft-dev
+ *
+ * MIT License
+ */
+
+package com.demonwav.mcdev.platform.architectury.version
+
+import com.demonwav.mcdev.util.SemanticVersion
+import com.intellij.util.castSafelyTo
+import com.jetbrains.rd.util.first
+import java.io.IOException
+import java.net.URL
+import javax.xml.stream.XMLInputFactory
+import javax.xml.stream.events.XMLEvent
+import kotlinx.serialization.json.Json
+import kotlinx.serialization.json.jsonObject
+import kotlinx.serialization.json.jsonPrimitive
+
+class ArchitecturyVersion private constructor(val versions: Map<SemanticVersion, List<SemanticVersion>>) {
+
+    fun getArchitecturyVersions(mcVersion: SemanticVersion): List<SemanticVersion> {
+        val roundedVersion = Json.parseToJsonElement(
+            URL(
+                "https://gist.githubusercontent.com" +
+                    "/shedaniel/4a37f350a6e49545347cb798dbfa72b3" +
+                    "/raw/architectury.json"
+            ).readText()
+        ).jsonObject["versions"]!!.jsonObject.map { SemanticVersion.parse(it.key) }
+            .windowed(2, 1).toMutableList().let {
+                it.add(
+                    listOf(
+                        it.last().last(),
+                        SemanticVersion.release(
+                            (
+                                it.last().last().take(2)
+                                    .parts[0]
+                                    .castSafelyTo<SemanticVersion.Companion.VersionPart.ReleasePart>()!!
+                                )
+                                .version + 1
+                        )
+                    )
+                ); it
+            }.find { mcVersion >= it[0] && mcVersion < it[1] }!!.first()
+        return versions[roundedVersion]!!.asSequence()
+            .sortedDescending()
+            .take(50)
+            .toList()
+    }
+
+    companion object {
+        private fun findMcVersion(architecturyVersion: SemanticVersion): SemanticVersion {
+            val meta = Json.parseToJsonElement(
+                URL(
+                    "https://gist.githubusercontent.com" +
+                        "/shedaniel/4a37f350a6e49545347cb798dbfa72b3" +
+                        "/raw/architectury.json"
+                ).readText()
+            ).jsonObject
+            return SemanticVersion.parse(
+                meta["versions"]!!.jsonObject.filter {
+                    it.value.jsonObject["api"]!!.jsonObject["filter"]!!.jsonPrimitive.content.toRegex().matches(
+                        architecturyVersion.toString()
+                    )
+                }.first().key
+            )
+        }
+        fun downloadData(): ArchitecturyVersion? {
```

Comment:
This function (and maybe some others in this file) could be coroutines, which would allow sending these get requests in parallel.

Reply:
technically downloadData() is called in 
https://github.com/minecraft-dev/MinecraftDev/blob/22f04db2502b5f14f8da406dfbe3a61af1967436/src/main/kotlin/platform/architectury/creator/ArchitecturyProjectSettingsWizard.kt#L256-L266
so it is already asynchronus

Reply:
currently this is what i have
```kotlin
class ArchitecturyVersion private constructor(
    val versions: Map<SemanticVersion, List<SemanticVersion>>,
    private val mcVersions: MutableList<List<SemanticVersion>>
) {


    fun getArchitecturyVersions(mcVersion: SemanticVersion): List<SemanticVersion> {
        val roundedVersion = mcVersions.find { mcVersion >= it[0] && mcVersion < it[1] }?.first()
        return try {
            versions[roundedVersion]?.asSequence()
                ?.sortedDescending()
                ?.take(50)
                ?.toList() ?: throw IOException("Could not find any architectury versions for $mcVersion")
        } catch (e: IOException) {
            e.printStackTrace()
            emptyList()
        }
    }

    companion object {
        private val updateUrl = URL(
            "https://gist.githubusercontent.com" +
                    "/shedaniel/4a37f350a6e49545347cb798dbfa72b3" +
                    "/raw/architectury.json"
        )

        fun downloadData(): ArchitecturyVersion? {
            try {
                val meta = Json.parseToJsonElement(updateUrl.readText())
                val versions = mutableMapOf<SemanticVersion, MutableList<SemanticVersion>>()
                val mcVersions = meta.jsonObject["versions"]?.jsonObject?.map { SemanticVersion.parse(it.key) }
                    ?.windowed(2, 1)?.toMutableList().also {
                        it?.add(
                            listOf(
                                it.last().last(),
                                SemanticVersion.parse(
                                    when (val part = it.last().last().parts.getOrNull(1)) {
                                        is SemanticVersion.Companion.VersionPart.ReleasePart -> (part.version + 1).toString()
                                        null -> "?"
                                        else -> part.versionString
                                    }
                                )
                            )
                        )
                    } ?: throw IOException("Could not find any minecraft versions")

                meta.jsonObject["versions"]?.jsonObject?.forEach {
                    val mcVersion = SemanticVersion.parse(it.key)
                    URL(
                        it.value.jsonObject["api"]?.jsonObject?.get("pom")?.jsonPrimitive?.content ?: throw IOException(
                            "Could not find pom for $mcVersion"
                        )
                    )
                        .openStream().use { stream ->
                            val inputFactory = XMLInputFactory.newInstance()

                            @Suppress("UNCHECKED_CAST")
                            val reader = inputFactory.createXMLEventReader(stream) as Iterator<XMLEvent>
                            for (event in reader) {
                                if (!event.isStartElement) {
                                    continue
                                }
                                val start = event.asStartElement()
                                val name = start.name.localPart
                                if (name != "version") {
                                    continue
                                }

                                val versionEvent = reader.next()
                                if (!versionEvent.isCharacters) {
                                    continue
                                }
                                val version = versionEvent.asCharacters().data
                                versions.getOrPut(mcVersion) { mutableListOf() }.add(SemanticVersion.parse(version))

                            }
                        }
                }

                return ArchitecturyVersion(versions, mcVersions)
            } catch (e: IOException) {
                e.printStackTrace()
                return null
            }
        }
    }
}
```

Reply:
It is asynchronous to the UI thread but that doesn't mean they're running in parallel to each other. Currently, you're sending a get request for each Minecraft version serially, where you have to wait the length of the ping for any response at all. If the requests are in parallel, in theory you only have to wait the length of the ping once, rather than once for each Minecraft version (which will get worse as more versions are added).